doc/specs/#674 - PowerShell cmdlets.md

```diff
@@ -0,0 +1,434 @@
+
+>Note: Naming things is hard, and one of the conventions of PowerShell is to make nouns singular.
+---
+
+### Client command `winget install`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Install-WinGetPackage
+```
+**Parameters**
+
+-Manifest
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Version
+
+-Source
+
+-Scope
+
+-Exact
+
+-Interactive
+
+-Silent
+>Note: we may switch the default to "silent" from "silent with progress" in winget.exe so there may be changes here.
+
+-Locale
+
+-Log
+
+-Override
+
+-Location
+
+---
+
+### Client command `winget show`
+
+**PowerShell Cmdlet**
+```PowerShell
+Get-WinGetPackageVersion
+```
+
+-Detail
+
+-All
+
+-Version
+
+**Examples**
+
+>Note: The current behavior for `winget show vscode` would look like the example below.
+
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -Detail
+```
+
+>Note: The current behavior for `winget show vscode --versions` would look like the example below.
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -All
+```
+
+---
+
+### Client command `winget source`
+
+**PowerShell Cmdlets**
+
+```PowerShell
+Add-WinGetSource
+```
+
+```PowerShell
+Remove-WinGetSource
+```
+
+```PowerShell
+Get-WinGetSource
+```
+
+```PowerShell
+Reset-WinGetSource
+```
+
+**Example**
+> Note: `winget source export` was added as a helper for adding approved sources to Group Policy. To get the same values with PowerShell the example blow would perform the same behavior.
+
+
+```PowerShell
+Get-WinGetSource | convertto-json -compress
+```
+
+---
+
+### Client command `winget search`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Find-WinGetPackage
+```
+
+**Parameters**
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Tag
+
+-Command
+
+-Source
+
+-Count
+
+-Exact
+
+---
+
+### Client command `winget list`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Get-WinGetPackage
+```
+**Parameters**
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Source
+
+-Tag
+
+-Command
+
+-Count
+
+-Exact
+
+---
+
+### Client command `winget upgrade`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Update-WinGetPackage
+```
+**Parameters**
+
+-Manifest  
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Version
+
+-Source
+
+-Exact
+
+-Interactive
+
+-Silent
+
+-Log
+
+-Override
+
+-Location
+
+-All
+
+---
+
+### Client command `winget uninstall`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Uninstall-WinGetPackage
+```
+
+**Parameters**
+
+-Manifest  
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Version
+
+-Source
+
+-Exact
+
+-Interactive
+
+-Silent
+
+-Log
+
+**Example**
+
+>Note: The example below could be executed to remove all packages with "xbox" in their name, or in other metadata fields provided in the manifest (assumes there is a match with the installed package in a configured source).
+
+```PowerShell
+Get-WingetPackage xbox | UninstallWingetPackage
+```
+
+---
+
+### Client command `winget hash`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Get-WinGetHash
```

Comment:
Kind of redundant. See [Get-FileHash](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/get-filehash?view=powershell-7.1)

Reply:
Is Get-FileHash doing the same thing as the proposed get-wingethash?

Reply:
`winget hash` just gives you the hash of a file, so I think so, yeah.

Reply:
BUt they are different objects returned.
So Get-FileHash returns a .NET object  like this:
```powershellconsole
PS C:\Foo> get-filehash .\2019-01-03_19-15-45.png | gm

   TypeName: Microsoft.PowerShell.Commands.FileHashInfo

Name        MemberType Definition
----        ---------- ----------
Equals      Method     bool Equals(System.Object obj)
GetHashCode Method     int GetHashCode()
GetType     Method     type GetType()
ToString    Method     string ToString()
Algorithm   Property   string Algorithm {get;set;}
Hash        Property   string Hash {get;set;}
Path        Property   string Path {get;set;}
```

whereas Winget Hash returns a string, like this:
PS C:\Foo> winget hash -f .\aaa-airport.txt
Sha256: 1fcfecf92e8917f6bbfaa2668ec17357cd32c2c419db229cc0435e55b1e25485
PS C:\Foo> winget hash -f .\aaa-airport.txt | gm

   TypeName: System.String

Name                 MemberType            Definition
----                 ----------            ----------
Clone                Method                System.Object Clone(), System.Object ICloneable.Clone()
CompareTo            Method                int CompareTo(System.Object value), int CompareTo(string strB), int ICompar…
....
```

So they are not the same objects and contain different properties.

Reply:
This was marked as resolved, but resolved with what decision?

`winget.exe hash -f .\file.txt` may return a string but how is that relevant when the same string can be accessed with:

```powershell
(Get-FileHash -LiteralPath .\file.txt -Algorithm SHA256).Hash
```

? The `.Hash` property of the `Microsoft.PowerShell.Commands.FileHashInfo` object that `Get-FileHash` returns contains the hash-string and if someone really wants to recreate the exact format that winget outputs they could do so:

```powershell
$hash = Get-FileHash -LiteralPath .\get-pip.py -Algorithm SHA256
"{0}: {1}" -f $hash.Algorithm, $hash.Hash.ToLower()
```

I agree with @megamorf here, it's redundant and the generic `Get-FileHash` cmdlet is preferred.

Reply:
I'm fine with removing the Windows Package Manager specific hash command if the community agrees with using Get-FileHash. It reduces the surface area, and uses the standard PowerShell tool based on what I'm reading here. It was originally added to the CLI to help author manifests. The same holds true for calculating the MSIX Signature SHA256. I'm not sure if we would be able to remove that, or if there is another PowerShell incantation to determine the signature hash.

Reply:
So long as Winget can accept the hash created with Get-FIleHash - seems a no brainer to remove it.  The approach should always be to not reinvent wheels wherever possible.  

It also removes the need to worry about how to convert it to a cmdlet, name wise! :-)

Reply:
@denelon As far as I know there is no built-in command to achieve the same as 
```
PS> winget hash -f example.msix -m
Sha256: 6bfa8f184e0afeb38dc7112e920d06df4704b4b1068968167c1972a97c1cc5ee
SignatureSha256: 6149697e47bdb4a29cbc07a62a79ac57c0d7007062a3514e63322d00be179d7c
```

One approach in PowerShell would be to extract the package and then extract the information:

```
PS> Get-FileHash -Path example.msix

Algorithm       Hash                                                                   Path
---------       ----                                                                   ----
SHA256          6BFA8F184E0AFEB38DC7112E920D06DF4704B4B1068968167C1972A97C1CC5EE       C:\Users\megamorf\Downloads\example.msix

# Expand archive expects .zip extension
PS> Rename-Item -Path example.msix -NewName example.zip
PS> Expand-Archive -Path example.zip -DestinationPath $env:TEMP\MsixExtracted
PS> Get-FileHash -Path $env:TEMP\MsixExtracted\AppxSignature.p7x -Algorithm SHA256

Algorithm       Hash                                                                   Path
---------       ----                                                                   ----
SHA256          6149697E47BDB4A29CBC07A62A79AC57C0D7007062A3514E63322D00BE179D7C       C:\Users\megamorf\AppData\Local\Temp\MsixExtracted\AppxSignature.p7x
```

This would be kind of like this here:
```powershell
function Get-AppxSignatureHash
{
    [CmdletBinding()]
    Param
    (
        [Parameter(Mandatory=$true)]
        $Path
    )

    Add-Type -AssemblyName System.IO.Compression.FileSystem
    $Guid = [System.Guid]::NewGuid()
    $OutputPath = Join-Path $env:TEMP $Guid

    try {
        $zip = [System.IO.Compression.ZipFile]::OpenRead($Path)
        $SigFile = $zip.Entries | Where-Object { $_.FullName -like "AppxSignature.p7x" }
        [System.IO.Compression.ZipFileExtensions]::ExtractToFile($SigFile, $OutputPath, $true)
        $zip.Dispose()

        Get-FileHash -Path $OutputPath -Algorithm SHA256
    }
    catch {
        [PSCustomObject]@{Hash=""}
    }
    finally {
        Remove-Item -Path $TempDir -ErrorAction SilentlyContinue
    }
}

$Path = "C:\Users\nem\megamorf\example.msix"

[PSCustomObject]@{
    Sha256 = (Get-FileHash -Path $Path -Algorithm SHA256).Hash.ToLower()
    SignatureSha256 = (Get-AppxSignatureHash -Path $Path -Verbose).Hash.ToLower()
} | Format-List


Sha256          : 6bfa8f184e0afeb38dc7112e920d06df4704b4b1068968167c1972a97c1cc5ee
SignatureSha256 : 6149697e47bdb4a29cbc07a62a79ac57c0d7007062a3514e63322d00be179d7c
```

Reply:
What role does ``winget hash`` play in the sceme of things? Does it _just_ get the hash of a wingetted program?  If I understand it, winget itself could just drop the hash command and tell folks to just use Get-FileHash?

Reply:
It was a way to get the file hash and the signature hash (for an MSIX package) to make manifest creation a bit easier. We now have wingetcreate for that so it's a bit redundant, and will likely be removed when wingetcreate is 1.0 stable.